bdk_core/src/tx_graph.rs

```diff
@@ -0,0 +1,83 @@
+use bitcoin::{OutPoint, Transaction, Txid};
+
+use crate::{alloc::vec::Vec, collections::*};
+
+#[derive(Clone, Debug, Default)]
+pub struct TxGraph {
+    txs: HashMap<Txid, Transaction>, // Transaction enum? {Full|Partial}, or separate field?
+    spends: BTreeMap<OutPoint, HashSet<Txid>>,
+    // TODO: Aux outputs.
+    // TODO: fn: Fee of tx
+}
+
+impl TxGraph {
+    /// The outputs from the transaction with id `txid` that have been spent.
+    pub fn outspend(&self, outpoint: &OutPoint) -> Option<&HashSet<Txid>> {
+        self.spends.get(outpoint)
+    }
+
+    /// Each item contains the output index and the txid that spent that output.
+    pub fn outspends(
+        &self,
+        txid: Txid,
+    ) -> impl DoubleEndedIterator<Item = (u32, &HashSet<Txid>)> + '_ {
+        let start = OutPoint { txid, vout: 0 };
+        let end = OutPoint {
+            txid,
+            vout: u32::MAX,
+        };
+        self.spends
+            .range(start..=end)
+            .map(|(outpoint, spends)| (outpoint.vout, spends))
+    }
+
+    /// Get transaction by txid.
+    pub fn tx(&self, txid: &Txid) -> Option<&Transaction> {
+        self.txs.get(txid)
+    }
+
+    /// Add transaction, returns true when [`TxGraph`] is updated.
+    pub fn insert_tx(&mut self, tx: &Transaction) -> bool {
+        let txid = tx.txid();
+
+        if self.txs.insert(txid, tx.clone()).is_some() {
+            return false;
+        }
+
+        tx.input
+            .iter()
+            .map(|txin| txin.previous_output)
+            .for_each(|outpoint| {
+                self.spends
+                    .entry(outpoint)
+                    .or_insert_with(HashSet::new)
+                    .insert(txid);
+            });
+
+        (0..tx.output.len() as u32)
+            .map(|vout| OutPoint { txid, vout })
+            .for_each(|outpoint| {
+                self.spends.entry(outpoint).or_default();
+            });
+
+        return true;
+    }
+
+    pub fn is_unspent(&self, outpoint: &OutPoint) -> Option<bool> {
+        self.spends.get(outpoint).map(|txids| txids.is_empty())
+    }
+
+    /// Return all txids of conflicting transactions to the given tx.
+    pub fn conflicting_txids(&self, tx: &Transaction, include_self: bool) -> Vec<Txid> {
```

Comment:
suggestions:

- I don't think we ever want to `include_self`.
- Return an `impl Iterator`
- Make it an interator of `(OutPoint, u32)` meaning "the input at this `vin` of the passed in tx conflicts with the output at this `outpoint`.

Reply:
Do you mean an iterator of `(u32, Txid)`?